app/services/foreman_bootdisk/iso_generator.rb

```diff
@@ -76,12 +95,34 @@ def self.generate(opts = {})
 
         if opts[:ipxe]
           ipxe_source_file = File.join(Setting[:bootdisk_ipxe_dir], 'ipxe.lkrn')
-          raise(Foreman::Exception, N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
+          raise Foreman::Exception.new(N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
 
           FileUtils.cp(ipxe_source_file, File.join(wd, 'build', 'ipxe'))
           File.open(File.join(wd, 'build', 'script'), 'w') { |file| file.write(opts[:ipxe]) }
         end
 
+        if opts[:grub]
+          FileUtils.mkdir_p(File.join(wd, 'build', 'EFI', 'BOOT'))
+          grub_cfg = File.join(wd, 'build', 'EFI', 'BOOT', 'grub.cfg')
+          File.open(grub_cfg, 'w') { |file| file.write(opts[:grub]) }
+          efibootimg = File.join(wd, 'build', 'efiboot.img')
+          system_or_exception("truncate -s 4M #{efibootimg}", N_('Creating new image failed, install truncate utility'))
+          system_or_exception("mkfs.msdos #{efibootimg}", N_('Failed to format the ESP image via mkfs.msdos'))
+          system_or_exception("mmd -i #{efibootimg} '::/EFI'", N_('Failed to create a directory within the ESP image'))
+          system_or_exception("mmd -i #{efibootimg} '::/EFI/BOOT'", N_('Failed to create a directory within the ESP image'))
+          {'grubx64.efi' => 'GRUBX64.EFI', 'shimx64.efi' => 'BOOTX64.EFI'}.each do |source_file, destination_file|
+            src = File.join(Setting[:bootdisk_grub2_dir], source_file)
+            dst = "::/EFI/BOOT/#{destination_file}"
+            raise(Foreman::Exception.new(N_('Ensure %{path} contains grubx64.efi and shimx64.efi or change path in settings'), path: src)) unless File.exist?(src)
```

Comment:
Damn, this one makes it a little bit more complex. What I am going to do is to use `/var/lib/tftpboot/grub2` and if it does not exists it would fall back to `/var/lib/foreman/bootdisk` path and recommend the user to do this in the exception:

Ensure %{path} contains grubx64.efi and shimx64.efi: install TFTP module or copy those files into /var/lib/foreman/bootdisk (Grub2 directory setting)